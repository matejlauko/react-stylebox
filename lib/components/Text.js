'use strict';

exports.__esModule = true;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _jsxFileName = 'src/components/Text.js';

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _Box = require('./Box');

var _Box2 = _interopRequireDefault(_Box);

require('../types');

var _compose = require('../compose');

var _compute = require('../compute');

var _utils = require('../utils');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

var fontSize = function fontSize(typography, size) {
  return typeof size === 'number' ? typography.fontSize(size) : size;
};

var computeSetup = {
  size: function size(val, _ref) {
    var typography = _ref.typography;
    return { fontSize: fontSize(typography, val) };
  },
  color: function (_color) {
    function color(_x, _x2) {
      return _color.apply(this, arguments);
    }

    color.toString = function () {
      return _color.toString();
    };

    return color;
  }(function (color, theme) {
    return theme.colors[color] || color;
  }),
  fontFamily: function fontFamily(val) {
    return val;
  },
  align: function align(val) {
    return { textAlign: val };
  },
  bold: function bold(val, theme) {
    return {
      fontWeight: val === true ? (0, _utils.get)(theme, 'text.bold', 'bold') : val === false ? 'normal' : val
    };
  },
  decoration: function decoration(val) {
    return { textDecoration: val };
  },
  italic: function italic(val) {
    return { fontStyle: val && 'italic' };
  },
  lineHeight: function lineHeight(val) {
    return val;
  }
};

var Text = function Text(_ref2, _ref3) {
  var theme = _ref3.theme;

  var as = _ref2.as,
      style = _ref2.style,
      props = _objectWithoutProperties(_ref2, ['as', 'style']);

  var _computeStyles = (0, _compute.computeStyles)(computeSetup)(props, theme),
      textStyles = _computeStyles[0],
      restProps = _computeStyles[1];

  var composedStyles = (0, _compose.composeStyles)(style, textStyles);

  return _react2.default.createElement(_Box2.default, _extends({ as: as || 'span', style: composedStyles }, restProps, {
    __source: {
      fileName: _jsxFileName,
      lineNumber: 46
    },
    __self: undefined
  }));
};

Text.displayName = 'Text';

Text.contextTypes = {
  theme: _propTypes2.default.object
};

exports.default = Text;