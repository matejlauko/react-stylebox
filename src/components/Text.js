// @flow
import React from 'react';
import PropTypes from 'prop-types';
import Box, { type BoxProps } from './Box';
import { type Theme, type Color, type Bold } from '../types';
import { composeStyles } from './compose';
import { computeStyles } from './compute';
import { get } from '../utils';

export type TextProps = BoxProps & {
  fontFamily?: string,
  size?: number,
  align?: 'left' | 'right' | 'center' | 'justify',
  bold?: Bold,
  color?: Color,
  decoration?: 'none' | 'underline' | 'line-through',
  italic?: boolean,
  lineHeight?: number,
};

type TextContext = {
  theme: Theme,
};

const fontSize = (typography, size) => {
  return typeof size === 'number' ? typography.fontSize(size) : size;
};

const computeSetup = {
  size: (val, { typography }) => ({ fontSize: fontSize(typography, val) }),
  color: (color, theme) => theme.colors[color] || color,
  fontFamily: val => val,
  align: val => ({ textAlign: val }),
  bold: (val, theme) => ({
    fontWeight: val === true ? get(theme, 'text.bold', 'bold') : val === false ? 'normal' : val,
  }),
  decoration: val => ({ textDecoration: val }),
  italic: val => ({ fontStyle: val && 'italic' }),
  lineHeight: val => val,
};

const Text = ({ as, style, ...props }: TextProps, { theme }: TextContext) => {
  const [textStyles, restProps] = computeStyles(computeSetup)(props, theme);
  const composedStyles = composeStyles(style, textStyles);

  return <Box as={as || 'span'} style={composedStyles} {...restProps} />;
};

Text.displayName = 'Text';

Text.contextTypes = {
  theme: PropTypes.object,
};

export default Text;
