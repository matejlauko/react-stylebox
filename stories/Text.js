import React from 'react';
import { storiesOf } from '@storybook/react';
import { Text, Heading } from '../src';

storiesOf('Text/Text', module)
  .add('styled w/ style', () => (
    <Text
      style={{
        color: 'red',
      }}
    >
      Text
    </Text>
  ))
  .add('styled w/ theme', () => (
    <Text
      style={theme => ({
        color: theme.colors.secondary,
      })}
    >
      Text
    </Text>
  ))
  .add('styled w/ computed props', () => (
    <Text
      size={2}
      italic
      fontFamily="Arial"
      color="primary"
      align="right"
      bold
      decoration="line-through"
      lineHeight={1.5}
    >
      Text
    </Text>
  ));

storiesOf('Text/Heading', module)
  .add('styled w/ style', () => <Heading>Heading</Heading>)
  .add('styled w/ theme', () => (
    <Heading
      style={theme => ({
        fontWeight: 'normal',
        color: theme.colors.secondary,
      })}
    >
      Heading
    </Heading>
  ))
  .add('styled w/ computed props', () => (
    <Heading bold={false} color="red" size={5} marginTop={3}>
      Heading
    </Heading>
  ))
  .add('styled w/ default size', () => <Heading type={3}>Heading 3</Heading>);
