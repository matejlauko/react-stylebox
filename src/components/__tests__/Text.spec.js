import React from 'react';
import renderFelaComponent from './render';
import Text from '../Text';

[{}].forEach(props => {
  it(`renders correctly with props, ${JSON.stringify(props, null, 2)}`, () => {
    expect(renderFelaComponent(<Text {...props} />)).toMatchSnapshot();
  });
});
