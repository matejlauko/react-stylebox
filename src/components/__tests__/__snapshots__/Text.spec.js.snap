// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly with props, {} 1`] = `
Object {
  "component": <span
    className="xe296pg"
  />,
  "styles": "
    .xe296pg{position:relative}",
}
`;
